app/src/main/java/org/mozilla/fenix/components/metrics/AppLaunchTimeMeasurement.kt

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.os.Process
+import android.os.SystemClock
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.HOT
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.WARM
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.COLD
+import org.mozilla.fenix.components.metrics.AppLaunchTimeMeasurement.StartupMode.UNKNOWN
+import org.mozilla.fenix.perf.Stat
+
+/**
+ * Handles the logic of figuring out launch time for cold, warm and hot startup.
+ */
+class AppLaunchTimeMeasurement(private val stats: Stat = Stat()) {
+
+    private var isOnPreDrawCalled = false
+    private var startupMode = UNKNOWN
+
+    private var applicationOnCreateTimeStampNanoSeconds: Long? = null
+    private var homeActivityInitTimeStampNanoSeconds: Long? = null
+    private var homeActivityOnRestartTimeStampNanoSeconds: Long? = null
+    // we are considering screen to be visible after the first pre draw call.
+    private var homeActivityOnPreDrawTimeStampNanoSeconds: Long = INVALID_TIMESTAMP
+
+    fun onApplicationOnCreate() {
+        startupMode = COLD
+    }
+
+    fun onHomeActivityOnCreate(activityInitNanos: Long) {
+        if (startupMode == UNKNOWN) {
+            startupMode = WARM
+        }
+        this.homeActivityInitTimeStampNanoSeconds = activityInitNanos
+    }
+
+    fun onHomeActivityOnRestart(activityOnRestartNanos: Long = SystemClock.elapsedRealtimeNanos()) {
+        startupMode = HOT
+        homeActivityOnRestartTimeStampNanoSeconds = activityOnRestartNanos
+    }
+
+    fun onActivityOnPreDraw(activityOnPreDrawNanos: Long = SystemClock.elapsedRealtimeNanos()) {
```

Comment:
nit: I wonder if we should rename this and the variables in this function to be `onFirstFrameDrawn` – that's what we're trying to represent with these values and it's not clear anywhere in this file right now.

Reply:
I guess its ok to do that as long as it is common to assume onPreDraw() to be same as firstFrameDrawn?. I always assumed first frame drawn = first call to onDraw. not sure how many MS diff it would be.

Reply:
That's true – maybe it's preferred to say `onFirstFramePreDraw` :). I think they're close enough to that it's not worth implementing a more complex solution, given onPreDraw is already implemented for us.

Reply:
done